File: ios/LivtetJigsaw/Sources/LivtetJigsaw/Model/Vector.swift
import CoreFoundation

public struct Vector: Equatable, Codable, AdditiveArithmetic {
    public let x: CGFloat
    public let y: CGFloat

    public init(x: CGFloat, y: CGFloat) {
        self.x = x
        self.y = y
    }

    public static let zero = Vector(x: 0, y: 0)

    public static func + (l: Vector, r: Vector) -> Vector {
        Vector(x: l.x + r.x, y: l.y + r.y)
    }

    public static func - (l: Vector, r: Vector) -> Vector {
        Vector(x: l.x - r.x, y: l.y - r.y)
    }

    public static func * (l: Vector, r: CGFloat) -> Vector {
        Vector(x: l.x * r, y: l.y * r)
    }

    public static func / (l: Vector, r: CGFloat) -> Vector {
        Vector(x: l.x / r, y: l.y / r)
    }

    public static func cast(_ v: CGFloat) -> Vector {
        Vector(x: v, y: v)
    }

    public static func cast(_ v: Int) -> Vector {
        let c = CGFloat(v)
        return Vector(x: c, y: c)
    }

    public static func plus(_ a: Vector, _ b: Vector) -> Vector {
        a + b
    }

    public static func minus(_ a: Vector, _ b: Vector) -> Vector {
        a - b
    }

    public static func multiply(_ a: Vector, _ b: Vector) -> Vector {
        Vector(x: a.x * b.x, y: a.y * b.y)
    }

    public static func copy(_ v: Vector) -> Vector {
        v
    }

    public static func distance(_ a: Vector, _ b: Vector) -> CGFloat {
        let dx = a.x - b.x
        let dy = a.y - b.y
        return sqrt(dx * dx + dy * dy)
    }

    public static func isClose(_ a: Vector, _ b: Vector, tolerance: CGFloat) -> Bool {
        abs(a.x - b.x) <= tolerance && abs(a.y - b.y) <= tolerance
    }
}

File: android/jigsaw/src/main/kotlin/net/olamaelcu/livtet/jigsaw/model/Vector.kt
package net.olamaelcu.livtet.jigsaw.model

import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

data class Vector(val x: Float, val y: Float) {
    companion object {
        val ZERO = Vector(0f, 0f)

        fun cast(value: Float): Vector = Vector(value, value)

        fun cast(value: Vector): Vector = value

        fun copy(v: Vector): Vector = Vector(v.x, v.y)

        fun equal(one: Vector, other: Vector, delta: Float = 0f): Boolean =
            abs(one.x - other.x) <= delta && abs(one.y - other.y) <= delta

        fun diff(one: Vector, other: Vector): Pair<Float, Float> =
            (one.x - other.x) to (one.y - other.y)

        fun multiply(one: Float, other: Float): Vector = apply(one, other) { a, b -> a * b }
        fun multiply(v: Vector, scalar: Float): Vector = Vector(v.x * scalar, v.y * scalar)
        fun multiply(v: Vector, other: Vector): Vector = Vector(v.x * other.x, v.y * other.y)

        fun divide(v: Vector, other: Vector): Vector = Vector(v.x / other.x, v.y / other.y)
        fun divide(v: Vector, scalar: Float): Vector = Vector(v.x / scalar, v.y / scalar)

        fun plus(v: Vector, other: Vector): Vector = Vector(v.x + other.x, v.y + other.y)
        fun minus(v: Vector, other: Vector): Vector = Vector(v.x - other.x, v.y - other.y)

        fun min(v: Vector, other: Vector): Vector = Vector(min(v.x, other.x), min(v.y, other.y))
        fun max(v: Vector, other: Vector): Vector = Vector(max(v.x, other.x), max(v.y, other.y))

        private fun apply(one: Float, other: Float, f: (Float, Float) -> Float): Vector {
            val first = cast(one)
            val second = cast(other)
            return Vector(f(first.x, second.x), f(first.y, second.y))
        }

        fun innerMin(v: Vector): Float = min(v.x, v.y)
        fun innerMax(v: Vector): Float = max(v.x, v.y)
    }
}
